src/Cis194/Hw/LogAnalysis.hs

```diff
@@ -6,22 +6,44 @@ module Cis194.Hw.LogAnalysis where
 import  Cis194.Hw.Log
 
 parseMessage :: String -> LogMessage
-parseMessage s = Unknown s
+parseMessage s =
+  case words s of
+    "E":i:ws -> buildMessage (Error (read i)) ws
+    "I":ws -> buildMessage Info ws
+    "W":ws -> buildMessage Warning ws
+    _ -> Unknown s
+
+buildMessage :: MessageType -> [String] -> LogMessage
+buildMessage m (t:ws) = LogMessage m (read t) (unwords ws)
 
 parse :: String -> [LogMessage]
-parse _ = []
+parse = map (parseMessage) . lines
 
 insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) t = t
+insert l Leaf = Node Leaf l Leaf
+insert n@(LogMessage _ nt _) (Node l m@(LogMessage _ mt _) r)
+  | nt < mt = Node (insert n l) m r
+  | nt > mt = Node l m (insert n r)
 insert _ t = t
 
 build :: [LogMessage] -> MessageTree
-build _ = Leaf
+build = foldr insert Leaf . reverse
 
 inOrder :: MessageTree -> [LogMessage]
-inOrder _ = []
+inOrder (Node l m r) = inOrder l ++ [m] ++ inOrder r
+inOrder Leaf = []
```

Comment:
This implements the algorithm as specified, but it is somewhat inefficient to left-nest `++` as each nested use must be deconstructed by the next `++`. A common pattern for this is:

```hs
inOrder :: MessageTree -> [LogMessage]
inOrder t = go t []
  where
    go (Node l m r) = inOrder l . (m:) . inOrder r
    go Leaf = id
```

This pattern is captured by the [dlist](http://hackage.haskell.org/package/dlist) package as well as the `ShowS` type in base.